### PR TITLE
box: add detailed invalid-key-part error

### PR DESCRIPTION
**Pull request by Lord-KA**
Add ER_EXACT_MATCH_DETAILED ("Invalid key part count in an exact match") error code, as old version wasn't informative enough in some cases. Now it displays space and index names too.

Refactor exac_key_validate() and exac_key_validate_nullable() to receive wider context to provide names for the error msg. Now they are specialised to primary key and user-defined key accordingly.

NO_DOC=error is not described in doc

### CODE REVIEW

**Comment by alyapunov** [comment]
As I understood you have fixed only a small part of the problem:
```
tarantool> box.space._space:get({1, 1})
---
- error: Invalid key part count in an exact match in space _space, index primary (expected
    1, got 2)
...

tarantool> box.space._space:select({1, 1})
---
- error: Invalid key part count (expected [0..1], got 2)
...

tarantool> box.space._space:select({'a'})
---
- error: 'Supplied key type of part 0 does not match index part type: expected unsigned'
...
```
I fear this patch introduces some chaos: in some places there will be details, in some - not..

**Comment by alyapunov** [comment]
I also think that we should fix all places with ER_EXACT_MATCH.

File: changelogs/unreleased/gh-7223-update-invalid-key-part-error.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+## feature/box
+
+* Added detailed version of "invalid key part count" error with current
+  space and index names (gh-7332).
```
Comment (alyapunov) [inline changelogs/unreleased/gh-7223-update-invalid-key-part-error.md:4]: It must be 7223

**Comment by alyapunov** [comment]
We discussed offline, it would be much better if we use stack diagnostics.

**Comment by Mons** [comment]
> As I understood you have fixed only a small part of the problem:

It is better to fix at least part of problems, than not to fix them at all and leave it in LTS release as is for several years.

**Comment by alyapunov** [comment]
`exac_key_validate` -> `exact_key_validate` in the commit message (twice).

**Comment by alyapunov** [comment]
> > As I understood you have fixed only a small part of the problem:
> 
> It is better to fix at least part of problems, than not to fix them at all and leave it in LTS release as is for several years.

It must be fixed a least for `select` too, not for `get` alone.

File: src/box/index.h [hunk 2]
```diff
@@ -459,23 +459,23 @@ key_validate(const struct index_def *index_def, enum iterator_type type,
 	     const char *key, uint32_t part_count);
 
 /**
- * Check that the supplied key is valid for a search in a unique
- * index (i.e. the key must be fully specified).
+ * Check that the supplied primary index key is valid for a search in
```
Comment (alyapunov) [inline src/box/index.h:462]: That is not necessary a primary index, so you should not change the comment.

File: src/box/index.h [hunk 3]
```diff
@@ -459,23 +459,23 @@ key_validate(const struct index_def *index_def, enum iterator_type type,
...
  * @retval -1 The key is invalid.
  */
 int
-exact_key_validate(struct key_def *key_def, const char *key,
+exact_key_validate(struct index_def *index_def, const char *key,
 		   uint32_t part_count);
 
 /**
- * Check that the supplied key is valid for representing iterator
+ * Check that the supplied user-defined key is valid for representing iterator
```
Comment (alyapunov) [inline src/box/index.h:472]: Can't catch why you changed the comment here. I'm not sure that the function is used only for user-defined key, perhaps it is or will be used for developer-defined keys.

File: src/box/session_settings.c [hunk 4]
```diff
@@ -343,7 +343,8 @@ session_settings_space_execute_update(struct space *space, struct txn *txn,
 	const char *new_key, *key = request->key;
 	uint32_t new_size, new_key_len, key_len = mp_decode_array(&key);
 	if (key_len == 0) {
-		diag_set(ClientError, ER_EXACT_MATCH, 1, 0);
+		diag_set(ClientError, ER_EXACT_MATCH_DETAILED,
+			 space_name(space), pk_def->name, 2, 0);
```
Comment (alyapunov) [inline src/box/session_settings.c:347]: Why did you change `1` to `2`? There must be only one part, not two. It's a pity that we don't have a test for it. Perhaps we should add one.

**Review by alyapunov — CHANGES_REQUESTED** [review]
Thanks for the patch, see several comments below.

Generally I don't like that we fix only ER_EXACT_MATCH error, leaving lots of similar error aside. But I can agree that it is better that nothing.

But the thing that bothers me most is that we create another error ER_EXACT_MATCH_DETAILED along with the original ER_EXACT_MATCH. There's no logical difference between them, only message. For a stupid example, what will we do later if we need the same error, but without known instance of a space, just with index? Create another error ER_EXACT_MATCH_LESS_DETAILED?
I think we should add an extra optional token (%s) to the existing error and fill it with space and index when they are known, leaving as empty string when they are not known. That is not perfect but more flexible solution.

**Comment by alyapunov** [comment]
Note also that when we finally use stacked errors, and these changes in `exact_key_validate` and `exact_key_validate_nullable` will be rolled back.

File: src/box/index.h (hunk 3 above)
Comment (Lord-KA) [reply src/box/index.h:472]: As we've discussed privately, I changed `exact_key_validate_nullable()` to check `index_def->cmp_def` that is according to description a user defined key.

```C
         /**
           * User-defined key definition, merged with the primary
           * key parts. Used by non-unique keys to uniquely identify
           * iterator position.
           */
          struct key_def *cmp_def;
```

File: src/box/session_settings.c (hunk 4 above)
Comment (Lord-KA) [reply src/box/session_settings.c:347]: Fixed.

File: src/box/index.h (hunk 3 above)
Comment (Lord-KA) [reply src/box/index.h:472]: Ok, we've discussed it once more and I reworded the comments.

```diff
 /**
- * Check that the supplied key is valid for a search in a unique
- * index (i.e. the key must be fully specified).
+ * Check that the index_def->key_def is valid for a search in
+ * a unique index (i.e. the key must be fully specified).
  * @retval 0  The key is valid.
  * @retval -1 The key is invalid.
  */
 int
-exact_key_validate(struct key_def *key_def, const char *key,
+exact_key_validate(struct index_def *index_def, const char *key,
                   uint32_t part_count);
 
 /**
- * Check that the supplied key is valid for representing iterator
+ * Check that the index_def->cmp_def is valid for representing iterator
  * position (i.e. the key must be fully specified, but nulls are allowed).
  * @retval 0  The key is valid.
  * @retval -1 The key is invalid.
  */
 int
-exact_key_validate_nullable(struct key_def *key_def, const char *key,
+exact_key_validate_nullable(struct index_def *index_def, const char *key,
```

File: src/box/index.h (hunk 2 above)
Comment (Lord-KA) [reply src/box/index.h:462]: Fixed. See diff below.

File: changelogs/unreleased/gh-7223-update-invalid-key-part-error.md (hunk 1 above)
Comment (Lord-KA) [reply changelogs/unreleased/gh-7223-update-invalid-key-part-error.md:4]: Fixed:

```md
## feature/box

* Updated "invalid key part count" error to include current space and index
  names when possible (gh-7223).
```

**Comment by Lord-KA** [comment]
Removed `DETAILED` version.

File: src/box/index.h [hunk 5]
```diff
@@ -459,23 +459,23 @@ key_validate(const struct index_def *index_def, enum iterator_type type,
 	     const char *key, uint32_t part_count);
 
 /**
- * Check that the supplied key is valid for a search in a unique
- * index (i.e. the key must be fully specified).
+ * Check that the index_def->key_def is valid for a search in
```
Comment (alyapunov) [inline src/box/index.h:462]: That's wrong again.
The supplied key - is `key` argument.
I think you should write something like:
```
 * Check that the supplied key is valid for a search in a unique
 * index, i.e. the key must be fully specified and correspond to
 * index_def->key_def.
```

File: src/box/index.h [hunk 6]
```diff
@@ -459,23 +459,23 @@ key_validate(const struct index_def *index_def, enum iterator_type type,
...
  * @retval -1 The key is invalid.
  */
 int
-exact_key_validate(struct key_def *key_def, const char *key,
+exact_key_validate(struct index_def *index_def, const char *key,
 		   uint32_t part_count);
 
 /**
- * Check that the supplied key is valid for representing iterator
+ * Check that the index_def->cmp_def is valid for representing iterator
```
Comment (alyapunov) [inline src/box/index.h:472]: That's wrong again.
The supplied key - is key argument.
I think you should write something like:
```
* Check that the supplied key is valid for representing iterator
* position, i.e. the key must be fully specified and correspond to
* index_def->cmp_def, with nulls are allowed.
* Note that since this function is designed checking a key for
* precise iterator position, it relies on index_def->cmp_def,
* in contrast to `key_validate` which relies on index_def->key_def.
* 
```

**Review by alyapunov — CHANGES_REQUESTED** [review]

File: src/box/index.h (hunk 6 above)
Comment (Lord-KA) [reply src/box/index.h]: Done.

File: src/box/index.h (hunk 5 above)
Comment (Lord-KA) [reply src/box/index.h]: Done.

**Review by alyapunov — APPROVED** [review]

File: src/box/index.cc [hunk 7]
```diff
@@ -142,27 +142,35 @@ key_validate(const struct index_def *index_def, enum iterator_type type,
...
 	assert(key != NULL || part_count == 0);
+	struct key_def *key_def = index_def->key_def;
 	if (key_def->part_count != part_count) {
-		diag_set(ClientError, ER_EXACT_MATCH, key_def->part_count,
-			 part_count);
+		struct space *space = space_by_id(index_def->space_id);
+		diag_set(ClientError, ER_EXACT_MATCH,
+			 tt_sprintf("in space \"%s\", index \"%s\" ",
+				    space_name(space), index_def->name),
+			 key_def->part_count, part_count);
```
Comment (locker) [inline src/box/index.cc:155]: 1. Updated the submodule in [Tarantool EE](https://github.com/tarantool/tarantool-ee) to this PR.

2. Applied the following patch to make the project compile.

   <details>

   ```diff
      diff --git a/src/box/index_read_view.c b/src/box/index_read_view.c
   index 7c6c562fef9a..d4dfa56b29f4 100644
   --- a/src/box/index_read_view.c
   +++ b/src/box/index_read_view.c
   @@ -190,7 +190,7 @@ box_index_read_view_get(struct index_read_view *rv, const char *key,
    		return -1;
    	}
    	uint32_t part_count = mp_decode_array(&key);
   -	if (exact_key_validate(rv->def->key_def, key, part_count) != 0)
   +	if (exact_key_validate(rv->def, key, part_count) != 0)
    		return -1;
    	if (index_read_view_get(rv, key, part_count, result) != 0)
    		return -1;
   diff --git a/src/box/raw_read_view.c b/src/box/raw_read_view.c
   index 4e83bbd8850f..a0038bf464a4 100644
   --- a/src/box/raw_read_view.c
   +++ b/src/box/raw_read_view.c
   @@ -318,7 +318,7 @@ box_raw_read_view_get(const box_raw_read_view_index_t *index_,
    		return -1;
    	}
    	uint32_t part_count = mp_decode_array(&key);
   -	if (exact_key_validate(index->def->key_def, key, part_count) != 0)
   +	if (exact_key_validate(index->def, key, part_count) != 0)
    		return -1;
    	return index_read_view_get_raw(index, key, part_count, data, size);
    }
   ```

   </details>

3. Ran the following script.

   <details>

   ```Lua
   box.cfg({log_level = 'warn'})
   box.schema.create_space('test')
   box.space.test:create_index('pk')
   rv = box.read_view.open()
   box.space.test:drop()
   rv.space.test:get({1, 2})
   ```

   </details>

4. Got a segmentation fault.

   <details>

   ```
   Segmentation fault
     code: SEGV_MAPERR
     addr: 0xc0
     context: 0x7facfedff540
     siginfo: 0x7facfedff670
     rax      0x0                0
     rbx      0x55ff7b7240a0     94555776106656
     rcx      0x14               20
     rdx      0x100000           1048576
     rsi      0x100000           1048576
     rdi      0x0                0
     rsp      0x7facfedffc00     140380987194368
     rbp      0x7facfedffc00     140380987194368
     r8       0x7facfedfffd0     140380987195344
     r9       0x0                0
     r10      0x7facff03b000     140380989534208
     r11      0x10               16
     r12      0x1                1
     r13      0x0                0
     r14      0x40dfc088         1088405640
     r15      0x55ff7a7bd4b0     94555759957168
     rip      0x55ff7a57bd07     94555757591815
     eflags   0x10246            66118
     cs       0x33               51
     gs       0x0                0
     fs       0x0                0
     cr2      0xc0               192
     err      0x4                4
     oldmask  0x0                0
     trapno   0xe                14
   Current time: 1676469433
   Please file a bug at https://github.com/tarantool/tarantool/issues
   Attempting backtrace... Note: since the server has already crashed,
   this may fail as well
   #1  0x55ff7a7f42d6 in crash_collect+256
   #2  0x55ff7a7f4ce8 in crash_signal_cb+100
   #3  0x7fad016da420 in __restore_rt+0
   #4  0x55ff7a57bd07 in space_name+12
   #5  0x55ff7a57cdc5 in exact_key_validate+257
   #6  0x55ff7a7b8365 in box_index_read_view_get+588
   #7  0x55ff7a7bd591 in lbox_index_read_view_get+225
   #8  0x55ff7a8589f7 in lj_BC_FUNCC+70
   #9  0x55ff7a863875 in lua_pcall+849
   #10 0x55ff7a7d2f78 in luaT_call+45
   #11 0x55ff7a7c8c58 in lua_main+189
   #12 0x55ff7a7c9200 in run_script_f+1423
   #13 0x55ff7a565cca in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+34
   #14 0x55ff7a80016d in fiber_loop+210
   #15 0x55ff7aaca17d in coro_init+120
   Aborted
   ```

   </details>

**Review by locker — CHANGES_REQUESTED** [review]

**Comment by Lord-KA** [comment]
It was decided to all error enhancements in one go via stack diagnostics.
